File: projects/rpn2tex/migrations_archive/rust-feature-by-feature-3/src/lexer.rs
//! Lexical analyzer for RPN expressions.

use crate::error::{Error, Result};
use crate::tokens::{Token, TokenType};

/// A lexer that tokenizes RPN input.
///
/// # Examples
///
/// ```
/// use rpn2tex::lexer::Lexer;
///
/// let lexer = Lexer::new("42");
/// let tokens = lexer.scan_tokens().unwrap();
/// assert_eq!(tokens.len(), 1);
/// ```
#[derive(Debug)]
pub struct Lexer {
    source: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
    column: usize,
}

impl Lexer {
    /// Creates a new lexer for the given source string.
    #[must_use]
    pub fn new(source: impl Into<String>) -> Self {
        Self {
            source: source.into(),
            tokens: Vec::new(),
            start: 0,
            current: 0,
            line: 1,
            column: 1,
        }
    }

    /// Scans the source and returns all tokens.
    ///
    /// # Errors
    ///
    /// Returns an error if an invalid character is encountered.
    pub fn scan_tokens(mut self) -> Result<Vec<Token>> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token()?;
        }
        Ok(self.tokens)
    }

    fn scan_token(&mut self) -> Result<()> {
        let c = self.advance();
        let start_line = self.line;
        let start_column = self.column - 1;

        match c {
            // Whitespace
            ' ' | '\r' | '\t' => Ok(()),
            '\n' => {
                self.line += 1;
                self.column = 1;
                Ok(())
            }
            // Numbers
            '0'..='9' => self.scan_number(c),
            // Operators
            '+' => {
                self.tokens
                    .push(Token::new(TokenType::Plus, "+", start_line, start_column));
                Ok(())
            }
            '-' => {
                // Check if this is a negative number (digit follows immediately)
                if !self.is_at_end() && self.peek().is_ascii_digit() {
                    // It's a negative number
                    self.scan_negative_number(start_line, start_column)
                } else {
                    // It's a subtraction operator
                    self.tokens
                        .push(Token::new(TokenType::Minus, "-", start_line, start_column));
                    Ok(())
                }
            }
            '*' => {
                self.tokens
                    .push(Token::new(TokenType::Mult, "*", start_line, start_column));
                Ok(())
            }
            '/' => {
                self.tokens
                    .push(Token::new(TokenType::Div, "/", start_line, start_column));
                Ok(())
            }
            _ => Err(Error::LexerError {
                message: format!("Unexpected character: '{c}'"),
                line: self.line,
                column: self.column - 1,
            }),
        }
    }

    fn scan_number(&mut self, first_digit: char) -> Result<()> {
        let start_line = self.line;
        let start_column = self.column - 1;
        let mut value = String::from(first_digit);

        // Scan integer part
        while !self.is_at_end() && self.peek().is_ascii_digit() {
            value.push(self.advance());
        }

        // Scan decimal part
        if !self.is_at_end() && self.peek() == '.' {
            value.push(self.advance());
            while !self.is_at_end() && self.peek().is_ascii_digit() {
                value.push(self.advance());
            }
        }

        self.tokens.push(Token::new(
            TokenType::Number,
            value,
            start_line,
            start_column,
        ));
        Ok(())
    }

    fn scan_negative_number(&mut self, start_line: usize, start_column: usize) -> Result<()> {
        let mut value = String::from("-");

        // Scan integer part
        while !self.is_at_end() && self.peek().is_ascii_digit() {
            value.push(self.advance());
        }

        // Scan decimal part
        if !self.is_at_end() && self.peek() == '.' {
            value.push(self.advance());
            while !self.is_at_end() && self.peek().is_ascii_digit() {
                value.push(self.advance());
            }
        }

        self.tokens.push(Token::new(
            TokenType::Number,
            value,
            start_line,
            start_column,
        ));
        Ok(())
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> char {
        let c = self.source.chars().nth(self.current).unwrap();
        self.current += c.len_utf8();
        self.column += 1;
        c
    }

    fn peek(&self) -> char {
        if self.is_at_end() {
            '\0'
        } else {
            self.source.chars().nth(self.current).unwrap()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scan_integer() {
        let lexer = Lexer::new("42");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "42");
        assert_eq!(tokens[0].line(), 1);
        assert_eq!(tokens[0].column(), 1);
    }

    #[test]
    fn test_scan_decimal() {
        let lexer = Lexer::new("3.14");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "3.14");
    }

    #[test]
    fn test_scan_with_whitespace() {
        let lexer = Lexer::new("  42  ");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].value(), "42");
    }

    #[test]
    fn test_scan_plus() {
        let lexer = Lexer::new("+");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Plus);
        assert_eq!(tokens[0].value(), "+");
    }

    #[test]
    fn test_scan_addition() {
        let lexer = Lexer::new("5 3 +");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 3);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "5");
        assert_eq!(tokens[1].token_type(), TokenType::Number);
        assert_eq!(tokens[1].value(), "3");
        assert_eq!(tokens[2].token_type(), TokenType::Plus);
        assert_eq!(tokens[2].value(), "+");
    }

    #[test]
    fn test_scan_minus() {
        let lexer = Lexer::new("-");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Minus);
        assert_eq!(tokens[0].value(), "-");
    }

    #[test]
    fn test_scan_subtraction() {
        let lexer = Lexer::new("5 3 -");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 3);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "5");
        assert_eq!(tokens[1].token_type(), TokenType::Number);
        assert_eq!(tokens[1].value(), "3");
        assert_eq!(tokens[2].token_type(), TokenType::Minus);
        assert_eq!(tokens[2].value(), "-");
    }

    #[test]
    fn test_scan_negative_number() {
        let lexer = Lexer::new("-5");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "-5");
    }

    #[test]
    fn test_scan_negative_decimal() {
        let lexer = Lexer::new("-3.14");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "-3.14");
    }

    #[test]
    fn test_disambiguation_minus_with_space() {
        // "5 - 3" should be: Number(5), Minus, Number(3)
        let lexer = Lexer::new("5 - 3");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 3);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "5");
        assert_eq!(tokens[1].token_type(), TokenType::Minus);
        assert_eq!(tokens[1].value(), "-");
        assert_eq!(tokens[2].token_type(), TokenType::Number);
        assert_eq!(tokens[2].value(), "3");
    }

    #[test]
    fn test_disambiguation_minus_no_space() {
        // "5-3" should be: Number(5), Number(-3)
        let lexer = Lexer::new("5-3");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 2);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "5");
        assert_eq!(tokens[1].token_type(), TokenType::Number);
        assert_eq!(tokens[1].value(), "-3");
    }

    #[test]
    fn test_scan_div() {
        let lexer = Lexer::new("/");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 1);
        assert_eq!(tokens[0].token_type(), TokenType::Div);
        assert_eq!(tokens[0].value(), "/");
    }

    #[test]
    fn test_scan_division() {
        let lexer = Lexer::new("10 2 /");
        let tokens = lexer.scan_tokens().unwrap();
        assert_eq!(tokens.len(), 3);
        assert_eq!(tokens[0].token_type(), TokenType::Number);
        assert_eq!(tokens[0].value(), "10");
        assert_eq!(tokens[1].token_type(), TokenType::Number);
        assert_eq!(tokens[1].value(), "2");
        assert_eq!(tokens[2].token_type(), TokenType::Div);
        assert_eq!(tokens[2].value(), "/");
    }
}

File: projects/rpn2tex/migrations/go-module-by-module-3/lexer.go
package rpn2tex

import (
	"fmt"
	"strings"
	"unicode"
)

// LexerError represents an error that occurs during lexical analysis.
type LexerError struct {
	Message string
	Line    int
	Column  int
}

// Error implements the error interface for LexerError.
func (e *LexerError) Error() string {
	return fmt.Sprintf("Line %d, column %d: %s", e.Line, e.Column, e.Message)
}

// Lexer performs lexical analysis on RPN input text.
type Lexer struct {
	text   string // Source text to tokenize
	pos    int    // Current position in text
	line   int    // Current line number (1-based)
	column int    // Current column number (1-based)
}

// NewLexer creates a new Lexer for the given input text.
func NewLexer(text string) *Lexer {
	return &Lexer{
		text:   text,
		pos:    0,
		line:   1,
		column: 1,
	}
}

// Tokenize scans the input text and returns a slice of all tokens.
// Returns an error if an unexpected character is encountered.
func (l *Lexer) Tokenize() ([]Token, error) {
	var tokens []Token

	for !l.atEnd() {
		l.skipWhitespace()
		if l.atEnd() {
			break
		}

		token, err := l.scanToken()
		if err != nil {
			return nil, err
		}
		tokens = append(tokens, token)
	}

	// Add EOF token at the end
	tokens = append(tokens, Token{
		Type:   EOF,
		Value:  "",
		Line:   l.line,
		Column: l.column,
	})

	return tokens, nil
}

// atEnd checks if we've reached the end of the input.
func (l *Lexer) atEnd() bool {
	return l.pos >= len(l.text)
}

// peek returns the current character without advancing.
func (l *Lexer) peek() rune {
	if l.atEnd() {
		return 0
	}
	return rune(l.text[l.pos])
}

// advance consumes the current character and returns it.
// Updates line and column tracking.
func (l *Lexer) advance() rune {
	if l.atEnd() {
		return 0
	}

	ch := rune(l.text[l.pos])
	l.pos++

	if ch == '\n' {
		l.line++
		l.column = 1
	} else {
		l.column++
	}

	return ch
}

// skipWhitespace skips over whitespace characters.
func (l *Lexer) skipWhitespace() {
	for !l.atEnd() {
		ch := l.peek()
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			l.advance()
		} else {
			break
		}
	}
}

// scanToken scans a single token from the current position.
func (l *Lexer) scanToken() (Token, error) {
	startLine := l.line
	startColumn := l.column

	ch := l.advance()

	// Check for operators
	switch ch {
	case '+':
		return Token{Type: PLUS, Value: "+", Line: startLine, Column: startColumn}, nil
	case '*':
		return Token{Type: MULT, Value: "*", Line: startLine, Column: startColumn}, nil
	case '/':
		return Token{Type: DIV, Value: "/", Line: startLine, Column: startColumn}, nil
	case '-':
		// Check if this is a negative number or subtraction operator
		// Negative number: '-' followed immediately by a digit (no space)
		if !l.atEnd() && unicode.IsDigit(l.peek()) {
			return l.scanNumber("-", startLine, startColumn)
		}
		return Token{Type: MINUS, Value: "-", Line: startLine, Column: startColumn}, nil
	}

	// Check for numbers
	if unicode.IsDigit(ch) {
		return l.scanNumber(string(ch), startLine, startColumn)
	}

	// Unsupported character
	return Token{}, &LexerError{
		Message: fmt.Sprintf("Unexpected character '%c'", ch),
		Line:    startLine,
		Column:  startColumn,
	}
}

// scanNumber scans a numeric literal (integer or decimal).
// prefix contains the already-consumed first character(s).
func (l *Lexer) scanNumber(prefix string, startLine, startColumn int) (Token, error) {
	var sb strings.Builder
	sb.WriteString(prefix)

	// Scan digits
	for !l.atEnd() {
		ch := l.peek()
		if unicode.IsDigit(ch) {
			sb.WriteRune(ch)
			l.advance()
		} else if ch == '.' {
			// Check if we've already seen a decimal point
			if strings.Contains(sb.String(), ".") {
				break
			}
			sb.WriteRune(ch)
			l.advance()
		} else {
			break
		}
	}

	return Token{
		Type:   NUMBER,
		Value:  sb.String(),
		Line:   startLine,
		Column: startColumn,
	}, nil
}
